### PR TITLE
add support to use the entire file

### PR DESCRIPTION
**Pull request by psifertex**
Fixes the problems described in #2.

I'm not sure whether just switching entirely to the new method is better (maybe not for large files?) so for now I just register two menu items, but it's a pretty quick implementation. Feel free to change any way you like or let me know if you have other preferences.

### CODE REVIEW

**Comment by paulfariello** [comment]
Thanks for the PR. I'm doing a quick review right now.
I agree on giving the user the choice of decompiling only a function or the full binary.

File: __init__.py [hunk 1]
```diff
@@ -30,10 +30,10 @@ def __init__(self, view, function):
 
         self.endianness = 'big' if self._view.endianness else 'little'
 
-        self.cmdline = ['retdec-decompiler.sh']
         self.cmdline.append('-m')
-        self.cmdline.append('raw')
+        self.cmdline.append(mode)
 
+        self.cmdline = ['retdec-decompiler.sh']
```
Comment (paulfariello) [inline __init__.py]: You must init cmdline before appending `'-m'` and `mode`. Better remove your last commit.

File: __init__.py [hunk 2]
```diff
@@ -43,6 +43,23 @@ def __init__(self, view, function):
...
+            raise Exception(err)
+
+        with open('{}.c'.format(filename), 'r') as f:
+            code = f.read()
+
+        os.unlink('{}.c'.format(filename))
+        os.unlink('{}.c.frontend.dsm'.format(filename))
+
+        code = self.merge_symbols(code)
+        self.render_output(code)
```
Comment (paulfariello) [inline __init__.py]: We shouldn't dup the code between the two `decompile_` functions.

File: __init__.py [hunk 3]
```diff
@@ -43,6 +43,23 @@ def __init__(self, view, function):
...
+        p = Popen(self.cmdline, stdout=PIPE, stderr=PIPE)
+        err = p.communicate()[1]
+        if err.startswith('Error'):
+            raise Exception(err)
+
+        with open('{}.c'.format(filename), 'r') as f:
+            code = f.read()
+
+        os.unlink('{}.c'.format(filename))
+        os.unlink('{}.c.frontend.dsm'.format(filename))
```
Comment (paulfariello) [inline __init__.py]: We must remove filename too.

File: __init__.py [hunk 4]
```diff
@@ -60,6 +77,38 @@ def decompile_function(self):
...
+            f.write(raw)
+
+        self.cmdline.append('--raw-section-vma')
+        self.cmdline.append('{:#x}'.format(start))
+
+        self.cmdline.append('--raw-entry-point')
+        self.cmdline.append('{:#x}'.format(start))
+
+        self.cmdline.append(filename)
+        return filename
```
Comment (paulfariello) [inline __init__.py]: This function is duplicated it appears twice. See just below.

File: __init__.py [hunk 5]
```diff
@@ -117,11 +166,19 @@ def render_output(self, code):
         colored_code = highlight(code, lexer, formatter)
         show_html_report('{}.c'.format(self._function.name), colored_code)
 
+def decompile_full(view, function):
+    try:
+        retdec = RetDec(view, function, "bin")
+        retdec.decompile_file()
+    except Exception as e:
+        log_error('failed to decompile function\n{}'.format(e.message))
```
Comment (paulfariello) [inline __init__.py]: Should fix the log error: `failed to decompile file`.

File: __init__.py [hunk 6]
```diff
@@ -60,6 +77,38 @@ def decompile_function(self):
...
+    def read_file(self):
+        start = self._function.start
+
+        #Doesn't just operate on the original file since in a .BNDB the original file may no longer exist
+        fd, filename = tempfile.mkstemp()
+        with os.fdopen(fd, 'wb') as f:
+            f.write(self._view.file.raw.read(0, len(self._view.file.raw)))
+
+        self.cmdline.append('--select-ranges')
+        self.cmdline.append('{:#x}-{:#x}'.format(start, start+1))
```
Comment (paulfariello) [inline __init__.py]: Don't know if this is the right way to pass the range. But I haven't tested it and it just seems odd.

**Review by paulfariello — CHANGES_REQUESTED** [review]
Thanks for the PR just a few minor things to fix and I'll merge it.

**Comment by paulfariello** [comment]
Could you rebase your PR onto develop ? I've added travis-ci and pylint automation.

File: __init__.py (hunk 6 above)
Comment (psifertex) [reply __init__.py]: Worked in testing and looks correct according to their examples: https://github.com/avast-tl/retdec/blob/master/scripts/retdec-decompiler.sh#L84

**Comment by psifertex** [comment]
Sorry about that, wasn't getting these notifications. I think I resolved all the issues. Squashed and rebased on dev.

One minor clarification -- I didn't change it to actually decompile the entire binary right now, it just passes the entire binary to retdec and specifies the specific function since otherwise a lot of important context is lost that breaks disassembly even for that function (see issue #2 for example). I just call it "full" or "slower" analysis for the time being, but it might be worth documenting in the readme what the differences are. 

Travis failed, but it doesn't look like anything I'm doing as far as I can tell?

File: __init__.py [hunk 7]
```diff
@@ -60,6 +78,20 @@ def decompile_function(self):
         code = self.merge_symbols(code)
         self.render_output(code)
 
+    def read_file(self):
+        start = self._function.start
+
+        #Doesn't just operate on the original file since in a .BNDB the original file may no longer exist
```
Comment (LunNova) [inline __init__.py]: This line and 167 are too long so it's failing pylint

**Review by paulfariello — APPROVED** [review]

**Comment by paulfariello** [comment]
I rebased your PR onto develop since there was some rewrite of the plugin to support retdec config file.
Only one question left, should we decompile on range `function.start, function.start+1` or on full function?

**Comment by psifertex** [comment]
Thanks. Sorry I was slow to get back to this. 

I used `select-ranges` because `select-function` required knowing the function name as opposed to an address, but specifying the start address of a specific function seems to result in just that function being decompiled, so believe that's the right choice but could be wrong. 

I don't think it would hurt to do start to start+1 (though I don't think it will help, either), but "full function" is a bit problematic. Mainly because there's no real definition of what that would be. If you chose the lowest address to the highest address it's possible to have interleaved basic blocks (most compilers don't do this, but there's no reason they couldn't) so another function not desired to be decompiled could also be decompiled.

**Pull request merged**